File: src/com/javascript-practice/binarytree/pathFinder.js
/**
 * Write a function, pathFinder, thats takes in the root of a binary tree and a target value.
 * The function should return an array representing a path to the target value.
 * if target value is not found in the tree, then return null.
 * 
 * you may assume that the tree containe unique values.
 * 
//      a
//    /   \
//   b     c
//  / \     \
// d   e     f

    pathFinder(a, 'e'); // -> [ 'a', 'b', 'e' ]
 */

    class Node {
        constructor(val) {
            this.val = val;
            this.left = null;
            this.right = null;
        }
    }

    const a = new Node("a");
    const b = new Node("b");
    const c = new Node("c");
    const d = new Node("d");
    const e = new Node("e");
    const f = new Node("f");
    
    a.left = b;
    a.right = c;
    b.left = d;
    b.right = e;
    c.right = f;


    const pathFinder = (root, target) => {
        const result = pathHelper(root, target);
        
        return result ? result.reverse() : null;
    }

    const pathHelper = (root, target) =>  {
        if(root == null) return null;

        if(target == root.val) return [root.val];
        const leftPath = pathHelper(root.left, target);
        if(leftPath!=null) {
            leftPath.push(root.val);
            return leftPath; // does not give the order we expect
        }
        const rightPath = pathHelper(root.right, target);
        if(rightPath!=null) {
            rightPath.push(root.val);
            return rightPath; // does not give the order we expect
        }
        return null;
    }


const result = pathFinder(a, 'e'); // -> [ 'a', 'b', 'e' ]
console.log(result);

File: src/com/structy/binaryTree/PathFinder.java
package com.structy.binaryTree;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class PathFinder {

    public List<String> pathFinder(Node root, String target){

        List<String> result  = helper(root, target, new ArrayList<>());
        if(result.size() > 0) {
            Collections.reverse(result);
        }
        return result;
    }
    private List<String> helper(Node root, String target, List<String> res){
        if(root == null)
            return null;
        if(target.equals(root.str)){
            res.add(root.str);
            return res;
        }
        List<String> left = helper(root.left, target, res);
        if(left!=null){
            left.add(root.str);
            return left;
        }
        List<String> right = helper(root.right, target, res);
        if(right!=null){
            right.add(root.str);
            return right;
        }
        return null;
    }

    public static void main(String[] args) {
        Node b = new Node("b");
        Node a = new Node("a");
        Node c = new Node("c");
        Node d = new Node("d");
        Node e = new Node("e");
        Node f = new Node("f");

        a.left = b;
        a.right = c;
        b.left = d;
        b.right = e;
        c.right = f;

        //      a
        //    /   \
        //   b     c
        //  / \     \
        // d   e     f
        PathFinder finder = new PathFinder();

        System.out.println(finder.pathFinder(a, "e")); // -> [ 'a', 'b', 'e' ]
    }
}
